handle N targets

hackergrrl opened the pull request:
In addition to handling N targets, I also made a few other changes while digging around:
- More standard/unixy USAGE format string
- Writes all usage & general errors to stderr rather than stdout
- URL validation (`url.Parse()` doesn't do validation)
- More comments throughout
- Returns `504 Gateway Timeout` when all targets fail (rather than `404 Not
  Found`)

Handling N targets ended up being a significant refactor, so it was difficult to separate that change from the above bullet pointed items. I can try and pull them all apart into separate commits, but they're all small things and shouldn't be too hard to examine inline.

hackergrrl committed:
Various goodies.

Standard USAGE format
Writes usage & general errors to stderr
**Handles N targets**
URL validation (url.Parse doesn't do validation)
Generally more comments throughout
Returns 504 Gateway Timeout when all targets fail (rather than 404 Not
Found)

whyrusleeping commented on main.go:
```diff
@@ -2,34 +2,60 @@ package main
+	failed := false
+	for i, v := range targets {
+		// Verify the target is HTTP or HTTPS.
+		if !strings.HasPrefix(v, "http://") && !strings.HasPrefix(v, "https://") {
```
i would rip this out into a separate function: `ValidateTarget` or something

whyrusleeping commented on main.go:
```diff
@@ -38,77 +64,79 @@ func main() {
 		go func() {
-			<-cancel_a
-			<-cancel_b
+			for i := range cancels {
```
this goroutine won't exit unless all requests get cancelled, need to close the channel for the selected request once we're done with it

hackergrrl replied:
Good catch! I found some other issues as a result of investigating this. (see commits)

hackergrrl committed:
Separate out url validation.

hackergrrl committed:
Handles failures better.

1. When 'failed' is closed !ok will be true -- we weren't properly
   hitting that case and the connection would stay open.
2. Cancels the request when a bad response code is hit.

whyrusleeping commented:
now that #1 is merged, give this one a rebase and i'll CR/M

hackergrrl committed:
indentation; handle all-request failure correctly

hackergrrl commented:
Updated. Please take another look @whyrusleeping!

whyrusleeping commented on main.go:
```diff
@@ -7,142 +7,150 @@ import (
 	"net/http"
 	"net/url"
 	"os"
+	"reflect"
```
Things that make me say "here we go..." for $200 Alex.

whyrusleeping commented on main.go:
```diff
@@ -7,142 +7,150 @@ import (
+	// Verify the target is HTTP or HTTPS.
+	if !strings.HasPrefix(u, "http://") && !strings.HasPrefix(u, "https://") {
+		return nil, fmt.Errorf("%s: must be an http(s) URL.\n", u)
+	}
```
style nitpick: spaces after closing brackets.

hackergrrl replied:
Not sure what this means. Are you saying you _want_ trailing spaces after the closing brackets?

(I assumed you meant _newline_, so I added those.)

whyrusleeping replied:
Like this:

``` go
if thing {
    return nil, error.Broken()
}

// comment
if nextThing {
```

whyrusleeping replied:
oh yeah, refreshed and saw it. thanks!

whyrusleeping commented on main.go:
```diff
@@ -7,142 +7,150 @@ import (
+	for i, v := range targets {
+		ua, err := validateUrl(v)
+		if err != nil {
+			fmt.Println(err.Error())
```
can just do:

``` go
fmt.Println(err)
```

hackergrrl committed:
use single channel for responses

hackergrrl commented on main.go:
```diff
@@ -8,141 +8,131 @@ import (
 	"net/url"
 	"os"
 	"strings"
+
```
Yeah, reflection isn't necessary here. Let's just use a single channel.

whyrusleeping commented on main.go:
```diff
@@ -7,142 +7,150 @@ import (
-				resp_b <- resp
+		// Wait for a successful response (or failures across the board).
+		for {
+			chosen, value, _ := reflect.Select(cases)
```
i really feel like you could get away without using reflection here. Have one channel that a goroutine can send its index on when its done or something, then can grab  the response from an array by its index.

whyrusleeping commented on main.go:
```diff
@@ -8,141 +8,131 @@ import (
+		ua, err := validateUrl(v)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			failed = true
```
we could just put the os.Exit in here

hackergrrl replied:
True -- my intent here was to print out **all** of the malformed URLs before bailing.

whyrusleeping commented on main.go:
```diff
@@ -8,141 +8,131 @@ import (
+		for i := range targets {
+			req := *r
+			req.URL.Host = urls[i].Host
+			req.URL, _ = url.Parse(r.URL.String())
```
is this necessary?

whyrusleeping replied:
oh wait. i see, the host part changes

whyrusleeping commented on main.go:
```diff
@@ -8,141 +8,131 @@ import (
+			go func() {
+				rt := &http.Transport{DisableKeepAlives: true}
+				res, err := rt.RoundTrip(&req)
+				if err != nil {
```
lets go with:

``` go
switch {
case err != nil:
...
case res.StatusCode >= 500 || res.StatusCode == 400:
...
default:
...
}
```

hackergrrl replied:
Oh hey, I really like this pattern! Go's `switch`es are really nice.

whyrusleeping commented on main.go:
```diff
@@ -8,141 +8,131 @@ import (
-			} else {
-				resp_b <- resp
+			for i := range fails {
+				<-fails[i]
```
this is back to the point where I don't think this goroutine is ever gonna exit. where do _all_ the fails get closed?

hackergrrl replied:
Ah, I see now. I had misunderstood how this piece worked -- corrected.

hackergrrl committed:
addresses concurrency bugs
